src/main/java/io/stargate/sgv2/jsonapi/service/embedding/operation/EmbeddingProvider.java

```diff
@@ -33,21 +84,67 @@ Uni<Response> vectorize(
    *
    * @return
    */
-  int maxBatchSize();
+  public abstract int maxBatchSize();
+
+  /**
+   * Helper method that has logic wrt whether OpenAI (azure or regular) accepts {@code "dimensions"}
+   * parameter or not.
+   *
+   * @param modelName OpenAI model to check
+   * @return True if given OpenAI model accepts (and expects} {@code "dimensions"} parameter; false
+   *     if not.
+   */
+  protected static boolean acceptsOpenAIDimensions(String modelName) {
+    return !modelName.endsWith("-ada-002");
+  }
+
+  /**
+   * Helper method to replace parameters in a template string with values from a map: placeholders
+   * are of form {@code {parameterName}} and matching value to look for in the map is String {@code
+   * "parameterName"}.
+   *
+   * @param template Template with placeholders to replace
+   * @param parameters Parameters to replace in the template
+   * @return Processed template with replaced parameters
+   */
+  protected String replaceParameters(String template, Map<String, Object> parameters) {
+    final int len = template.length();
+    StringBuilder baseUrl = new StringBuilder(len);
+
+    for (int i = 0; i < len; ) {
+      char c = template.charAt(i++);
+      int end;
+
+      if ((c != '{') || (end = template.indexOf('}', i)) < 0) {
+        baseUrl.append(c);
+        continue;
+      }
+      String key = template.substring(i, end);
+      i = end + 1;
+
+      Object value = parameters.get(key);
+      if (value == null) {
+        throw new IllegalArgumentException(
+            "Missing URL parameter '" + key + "' (available: " + parameters.keySet() + ")");
+      }
+      baseUrl.append(value);
+    }
+    return baseUrl.toString();
+  }
 
   /**
    * Record to hold the batchId and embedding vectors
    *
    * @param batchId - Sequence number for the batch to order the vectors.
    * @param embeddings - Embedding vectors for the given text inputs.
    */
-  record Response(int batchId, List<float[]> embeddings) {
+  public record Response(int batchId, List<float[]> embeddings) {
```

Comment:
Can also be `static`? (in interfaces, public and static are implied by default I think)

Reply:
I got a warning message `Modifier 'static' is redundant for inner records ` from IDE, probably no need to add it?

Reply:
Ok. Wasn't sure, then it's same as within an `interface`. So no need to add.